### PR TITLE
Enable lock in MarkdownEditor

### PR DESCRIPTION
**Pull request by asmsuechan**
Hi, I added a function which keeps Editor mode even though I move to another window.

pic

refs: https://github.com/BoostIO/Boostnote/issues/272

### CODE REVIEW

**Comment by kazup01** [comment]
LGTM
Great <3

File: browser/components/MarkdownEditor.js [hunk 1]
```diff
@@ -13,12 +14,18 @@ class MarkdownEditor extends React.Component {
     this.state = {
       status: 'PREVIEW',
       renderValue: props.value,
-      keyPressed: {}
+      keyPressed: {},
+      locked: false
```
Comment (asmsuechan) [inline browser/components/MarkdownEditor.js]: This is duplicated and not connected state with https://github.com/asmsuechan/Boostnote/blob/74ee6ae6ce047ae7c349b35ae048ae14930c547e/browser/main/Detail/MarkdownNoteDetail.js#L29.

Comment (asmsuechan) [inline browser/components/MarkdownEditor.js]: However, I'm not sure what I should connect them.

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 2]
```diff
@@ -200,6 +201,16 @@ class MarkdownNoteDetail extends React.Component {
...
   }
 
+  handleLockButtonClick () {
+    ee.emit('editor:lock')
+    this.focus()
+    this.setState({ locked: !this.state.locked })
+  }
+
+  toggleLockButton () {
+    return this.state.locked ? 'fa-lock' : 'fa-unlock-alt'
```
Comment (asmsuechan) [inline browser/main/Detail/MarkdownNoteDetail.js]: This does not collaborate with actual state of MarkdownEditor.

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 3]
```diff
@@ -235,6 +246,20 @@ class MarkdownNoteDetail extends React.Component {
             />
           </div>
           <div styleName='info-right'>
+            {(() => {
+              // TODO: get a state of MarkdownEditor somehow
+              const editorStatus='CODE'
+              let faClassName=`fa ${this.toggleLockButton()}`
```
Comment (asmsuechan) [inline browser/main/Detail/MarkdownNoteDetail.js]: I wanna get the current status PREVIEW/CODE of the MarkdownEditor.

**Comment by asmsuechan** [comment]
Current status is WIP because the lock/unlock icon is shown even PREVIEW mode.

**Comment by asmsuechan** [comment]
pic

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 4]
```diff
@@ -25,15 +25,23 @@ class MarkdownNoteDetail extends React.Component {
...
         title: '',
         content: ''
-      }, props.note)
+      }, props.note),
+      editorStatus: false,
+      isLocked: false
     }
     this.dispatchTimer = null
+
+    this.showLockButton = () => this.handleShowLockButton()
```
Comment (sota1235) [inline browser/main/Detail/MarkdownNoteDetail.js]: Please use `Function.prototype.bind`. It is more clearly which `this` is assigned.

```javascript
this.showLockButton = this.showLockButton.bind(this);
```

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 5]
```diff
@@ -200,10 +212,28 @@ class MarkdownNoteDetail extends React.Component {
     }
   }
 
+  handleLockButtonMouseDown (e) {
+    e.preventDefault()
+    ee.emit('editor:lock')
+    this.setState({ isLocked: !this.state.isLocked })
+  }
+
+  toggleLockButton () {
```
Comment (sota1235) [inline browser/main/Detail/MarkdownNoteDetail.js]: I think `getToggleLockButtonClassName` is better.

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 6]
```diff
@@ -235,6 +265,19 @@ class MarkdownNoteDetail extends React.Component {
             />
           </div>
           <div styleName='info-right'>
+            {(() => {
+              let faClassName=`fa ${this.toggleLockButton()}`
```
Comment (sota1235) [inline browser/main/Detail/MarkdownNoteDetail.js]: Can you use `const`?

File: browser/main/Detail/MarkdownNoteDetail.js [hunk 7]
```diff
@@ -235,6 +265,19 @@ class MarkdownNoteDetail extends React.Component {
...
+              if (this.state.editorStatus === 'CODE') {
+                return(
+                  <button styleName='info-right-button'
+                    onFocus={(e) => this.handleFocus(e)}
+                    onMouseDown={(e) => this.handleLockButtonMouseDown(e)}
+                  >
+                    <i className={faClassName}/>
+                  </button>
+                )
+              }
```
Comment (sota1235) [inline browser/main/Detail/MarkdownNoteDetail.js]: Writing this function before `return` is more readable.

```javascript
const lockButtonComponent =
  <button styleName='info-right-button'
    onFocus={(e) => this.handleFocus(e)}
    onMouseDown={(e) => this.handleLockButtonMouseDown(e)}
  >
    <i className={`fa ${this.toggleLockButton()}`}/>
  </button>;

/** */

return (
  {this.state.editorStatus === 'CODE' ? lockButtonComponent : '' }
);
```

**Comment by asmsuechan** [comment]
I fixed them.

File: browser/main/Detail/MarkdownNoteDetail.js (hunk 4 above)
Comment (asmsuechan) [inline browser/main/Detail/MarkdownNoteDetail.js]: We need to refactor other same style lines.

Comment (sota1235) [inline browser/main/Detail/MarkdownNoteDetail.js]: I know, so we should try to refactor little by little...

**Comment by kazup01** [comment]
I think Tooltip is needed when hover on lock icon. How about it?

**Comment by asmsuechan** [comment]
@kazup01 So do I.

**Pull request merged**